lib/src/revset.rs

```diff
@@ -41,8 +42,14 @@ use crate::store::Store;
 /// Error occurred during symbol resolution.
 #[derive(Debug, Error)]
 pub enum RevsetResolutionError {
-    #[error("Revision \"{0}\" doesn't exist")]
-    NoSuchRevision(String),
+    #[error("Revision \"{name}\" doesn't exist{}", match candidates.first() {
+        Some(candidate) => format!("; did you mean \"{candidate}\"?"),
+        None => "".to_string(),
```

Comment:
I would move formatting to the CLI layer to colorize it as a hint. See:

https://github.com/martinvonz/jj/blob/d82266887c86fe308cbd632b1ca862cf87d6ba10/src/cli_util.rs#L277-L285

I made it test string similarity at later pass, but I don't know if that's better than narrowing candidates by revset engine.

Reply:
Perfect, thanks for the hint 😉.

It looks like this now:

<img width="526" alt="Screenshot 2023-06-04 at 23 32 27" src="https://github.com/martinvonz/jj/assets/454057/d3530d36-eab6-44a7-8773-52072f15bf30">

> I made it test string similarity at later pass, but I don't know if that's better than narrowing candidates by revset engine.

- As per @ilyagr's comment, we might want to suggest things like alternative remotes, which might not pass the filter later, so we should delegate candidate selection to the revset engine.
- You could also imagine some kind of cloud monorepo revset backend which provides suggestions but doesn't want to return a huge list of candidate branches. (Or it could suggest commit hashes, etc.)
-Also, it makes it easier to unit test the candidate selection.

Reply:
Fair enough.

FWIW, for (ambiguous) commit hashes, it might be better to issue a separate query instead of collecting candidates upfront. There might be tons of prefix matches, but limiting them up to arbitrary number wouldn't make much sense.